File: py/src/aoc2020/math.py
import math


def egcd(a, b):
    if not b:
        return 1, 0, a
    q, r = a // b, a % b
    s, t, g = egcd(b, r)
    return t, s - q * t, g


def crt(r1, q1, r2, q2):
    q3 = math.lcm(q1, q2)
    t, _, g = egcd(q1 + q2, q3)
    r3 = (r1 * q2 + r2 * q1) * t
    assert not r3 % g
    r3 = r3 // g % q3
    if (r3 < 0) != (q3 < 0):
        r3 += q3
    return r3, q3

File: kt/src/main/kotlin/io/github/ephemient/aoc2020/math.kt
package io.github.ephemient.aoc2020

import java.math.BigInteger
import java.math.BigInteger.ONE
import java.math.BigInteger.ZERO

fun egcd(a: BigInteger, b: BigInteger): Triple<BigInteger, BigInteger, BigInteger> {
    if (b == ZERO) {
        return Triple(ONE, ZERO, a)
    }
    val (q, r) = a.divideAndRemainder(b)
    val (s, t, g) = egcd(b, r)
    return Triple(t, s - q * t, g)
}

fun crt(
    r1: BigInteger,
    q1: BigInteger,
    r2: BigInteger,
    q2: BigInteger
): Pair<BigInteger, BigInteger> {
    val q3 = q1 / q1.gcd(q2) * q2
    val (t, _, g) = egcd(q1 + q2, q3)
    var (r3, zero) = ((r1 * q2 + r2 * q1) * t).divideAndRemainder(g)
    check(zero == ZERO)
    r3 %= q3
    if (r3.signum() * q3.signum() < 0) r3 += q3
    return Pair(r3, q3)
}
